### ISSUE TITLE
Does a Model take into account its Collection.allow?

### ISSUE THREAD

**Issue by adriano-di-giovanni** [jagi/meteor-astronomy#203]
I have a local collection

```javascript
BalanceCollection = new Mongo.Collection(null);

BalanceCollection.allow({
  insert: function () {
    return true;
  },
  update: function () {
    return true;
  },
  remove: function () {
    return true;
  },
});
```

and an Astronomy model

```javascript
BalanceModel = Astro.Class({
  name: 'Balance',
  collection: BalanceCollection,
  transform: null,
  fields: {
    issuerId: {
      type: 'string',
      validator: Validators.required(),
    },
    holderId: {
      type: 'string',
      validator: Validators.required(),
    },
    balance: {
      type: 'number',
      'default': 0,
      validator: Validators.and([
        Validators.required(),
        Validators.gte(0),
      ]),
    },
  },
});
```

I also have a Tracker

```javascript
if (Meteor.isClient) {
  Tracker.autorun(function () {
    let userId = Meteor.userId();
    let voucherArray = VoucherCollection.find({ issuerId: userId }).fetch();
    _(voucherArray).each(function (voucher) {
      let selector = {
        issuerId: voucher.issuerId,
        holderId: voucher.holderId,
      };
      let modifier = {
        $inc: {
          balance: voucher.value,
        },
      };
      BalanceModel.upsert(selector, modifier);
    });
  });  
}
```

Despite the fact that BalanceCollection allows any operation, I get the client error `Not permitted. Untrusted code may only update documents by ID`.

Everything works if I replace `BalanceModel.upsert(selector, modifier)` with `BalanceCollection.upsert(selector, modifier)`.

My question is: does a Model take into account its Collection allow?

Thanks in advance,
Adriano

**Comment by shadowRR** [comment]
As stated by Meteor docs, you cannot use mongo.update() on the client, with anything else than an _id as a selector. As for mongo.upsert, it probably only works for now because you have the 'insecure' package installed.

*"Untrusted code can only modify a single document at once, specified by its _id. The modification is allowed only after checking any applicable allow and deny rules. The number of affected documents will be returned to the callback. Untrusted code cannot perform upserts, except in insecure mode."*

Source : http://docs.meteor.com/#/full/update

**Comment by adriano-di-giovanni** [comment]
Hi @shadowRR,
and thanks for your reply.

You're perfectly right but everything works if I call `BalanceCollection.upsert(selector, modifier)`.
`BalanceCollection` is a local collection allowing any operation.

It doesn't work if I call `BalanceModel.upsert(selector, modifier)`.
`BalanceModel` is an Astronomy model.

Behavior seems to be inconsistent.

**Comment by jagi** [comment]
`Class.upsert` propagates method execution to `Collection.upsert` but just add a little bit of Astronomy checking. So yes it takes allow/deny rules into account but it's not a source of your problem. As @shadowRR stated you can't update/upsert on the client with selectors that modifies more than one document. You have to provide the `_id` of a document being modified. As I see, your selector does do that:

```js
let selector = {
  issuerId: voucher.issuerId,
  holderId: voucher.holderId,
};
```

I'm sure that it's should work properly if we are talking about the `update` operation but in the `upsert` operation I'm also checking if a selector is for a single document. Maybe Meteor doesn't do it for `Collection.upsert` and I'm a little bit to strict here.

**Comment by adriano-di-giovanni** [comment]
I think it should be possible to let client perform any kind of operation on local collections. Maybe Meteor guys think it, too. That's why `BalanceCollection.upsert` works. It seems to me that your additional check can be the source of this behavior inconsistency. Does it make sense to you?

**Comment by jagi** [comment]
Oh ok I have a lot on my mind and didn't noticed that you are talking about local collection. Yep it should do everything without checking as it's a local collection. Agree in 100% :). So yep it's a bug. I will fix it.

**Comment by jagi** [comment]
Fixed in v1.2.4
